test/controllers/products_controller_test.rb

```diff
@@ -31,4 +41,86 @@
       expect(flash[:error]).must_equal "Unknown product"
     end
   end
+
+  describe "new" do
+    describe "logged in merchant" do
+      it "succeeds" do
+        perform_login
+        get new_product_path
+
+        must_respond_with :success
+      end
+    end
+    describe "not logged in" do
+      it "redirects user" do
+        get new_product_path
+
+        must_respond_with :redirect
+        expect(flash[:error]).must_equal "You must be logged in to add a new product"
+      end
+    end
+  end
+
+  describe "create" do
+    describe "logged in user" do
+      it "creates a product with valid data" do
+        merchant = perform_login
+        new_product = {product: {name: "Something amazing", merchant_id: merchant.id, price: 1000}}
+        expect {
+          post products_path, params: new_product
+        }.must_change "Product.count", 1
+
+        new_product_id = Product.find_by(name: "Something amazing").id
+
+        must_respond_with :redirect
+        must_redirect_to product_path(new_product_id)
+      end
+    end
+
```

Comment:
I see you have a test for creating a new product and it's passing. When I create a new product though, I get a 400 Bad Request in the terminal after the POST request and my new product does not show on the products page. Can we look at this together?

Reply:
also same. looks like prob with merchant? getting flash error 
> Could not add new product: {:merchant=>["must exist"]}